tests/test_parser.py

```diff
@@ -131,3 +134,121 @@ def test_string_utf8():
+        A = Select([], recursive=True)
+        A.values = [A, ""]
+        parser = EarleyCommitParser(A)
+        # Test that computing the parse tree doesn't hang
+        parser.parse_tree()
+        # Test that getting captures doesn't hang
+        parser.get_captures()
+
+    @pytest.mark.timeout(5)
+    def test_no_infinite_loop_with_terminal(self):
+        """
+        A -> A B
+        A ->
+        B -> 'x'
+        B ->
+
+        Loop occurs because `A -> A B` is a nullable rule
+        """
+        B = select(["x", ""])
+        A = select([B, ""], recurse=True)
+        parser = EarleyCommitParser(A)
+        # Test that computing the parse tree doesn't hang
+        parser.parse_tree()
+        # Test that getting captures doesn't hang
+        parser.get_captures()
+
+    @pytest.mark.timeout(5)
+    def test_no_infinite_loop_extra_indirection(self):
+        """
+        A -> A C
+        A -> B
+        A ->
+        B -> A
+        C -> 'x'
+
+        Loop occurs because `A -> B`, `B -> A` are nullable rules
+        """
+        C = Join(["x"])
+        # Initialize as nullable -- quirk in how nullability is determined in Select
+        B = Select([""])
+        # Initialize as nullable -- quirk in how nullability is determined in Select
+        A = Select([""])
+        B.values = [A]
+        A.values = [A + C, B, ""]
+        assert A.nullable
+        assert B.nullable
+        parser = EarleyCommitParser(A)
+        # Test that computing the parse tree doesn't hang
+        parser.parse_tree()
+        # Test that getting captures doesn't hang
+        parser.get_captures()
+
+    @pytest.mark.timeout(5)
+    def test_captures_from_root(self):
+        B = select(["x", ""], name="B")
+        A = select([B, ""], recurse=True, name="A")
+        parser = EarleyCommitParser(A)
+
+        with patch.object(
+            parser,
+            "_record_captures_from_root",
+            wraps=parser._record_captures_from_root,
+        ) as mock:
+            parser.consume_byte(b"x")
+            captures, _ = parser.get_captures()
+            assert mock.call_count == 1
+            assert captures == {"B": b"x", "A": b"x"}
+
+            parser.consume_byte(b"x")
+            captures, _ = parser.get_captures()
+            assert mock.call_count == 2
+            assert captures == {"B": b"x", "A": b"xx"}
+
+    @pytest.mark.timeout(5)
+    def test_partial_captures(self):
+        B = select(["x", ""], name="B")
+        A = select([B, ""], recurse=True, name="A")
+        C = A + "y"
+        parser = EarleyCommitParser(C)
+
+        with patch.object(
+            parser,
+            "_record_captures_partial",
+            wraps=parser._record_captures_partial,
+        ) as mock:
+            parser.consume_byte(b"x")
+            captures, _ = parser.get_captures()
+            assert mock.call_count == 1
+            assert captures == {"B": b"", "A": b"x"}
```

Comment:
Quick note here: I'm not sure what the semantics of partial captures are supposed to be, but this test is consistent with the existing behavior without these nullable loops